### PR TITLE
Replace blocking calls in ExpandCollapseSearchResponseListener by asynchronous requests

### LINKED ISSUE #23048: ExpandCollapseSearchResponseListener potentially executes blocking calls on a network thread
We use ExpandCollapseSearchResponseListener for field collapsing but this listener executes heavy operations in a blocking fashion. This could happen on a transport thread. We should do this in an async way.

I just spoke with @jimczi how to fix it and he is taking it over. For now I will put a fix into 5.x and master since this triggers quite often due to a recent change.

Note: this code is unreleased

### PR DESCRIPTION
**Pull request by jimczi**
This changes removes the SearchResponseListener that was used by the ExpandCollapseSearchResponseListener to expand collapsed hits.
Instead the expand calls are done directly in the SearchPhaseController in a asynchronous fashion.
The removal of SearchResponseListener is not a breaking change because it was never release.
This functionnality is not very useful as is and should be rethink during the great search refactoring that Simon is doing.

Closes #23048

### CODE REVIEW

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by jimczi** [comment]
Sorry @s1monw but I made a small change that needs another review. So instead of sending all the expand queries at once the last commit fires one query at a time. I think it's safer this way. Can you take a look ?

**Comment by jimczi** [comment]
I pushed another commit that uses a `multi_search` request to retrieve the `inner_hits` per group.
The maximum number of concurrent requests can be controlled per collapse request using `max_concurrent_group_searches` param. ^^ @s1monw, WDYT ?

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 1]
```diff
@@ -82,11 +83,17 @@
...
 
-    public SearchPhaseController(Settings settings, BigArrays bigArrays, ScriptService scriptService) {
+    public SearchPhaseController(Settings settings, BigArrays bigArrays, ScriptService scriptService, Client client) {
         super(settings);
         this.bigArrays = bigArrays;
         this.scriptService = scriptService;
+        this.client = client;
+    }
+
+    public Client getClient() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: hmm any chance we can add a method to `SearchTransportService` that fires off a multisearch via transport service directly? I'd love to not go through a client here?

**Comment by s1monw** [comment]
this looks awesome! I left one comment

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java (hunk 1 above)
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: Right, we don't need the client here.
Does https://github.com/elastic/elasticsearch/pull/23053/commits/2c34484c607926d39ec6f0152e84f6185eaa7fed works for you ?

**Review by s1monw — APPROVED** [review]
LGTM thanks for doing this

**Comment by jimczi** [comment]
Thanks for spotting this @s1monw.
Now merging to 5.x and 5.3

**Pull request merged**